packages/flutter/lib/src/services/restoration.dart

```diff
@@ -0,0 +1,893 @@
+  /// restoration data has been provided to the [RestorationManager] to restore
+  /// the application to a different state. In response to the notification,
+  /// listeners must stop using the old root bucket and obtain the new one via
+  /// this getter ([rootBucket] will have been updated to return the new bucket
+  /// just before the old root bucket informs its listeners).
+  ///
+  /// The restoration data describing the current bucket hierarchy is retrieved
+  /// asynchronously from the engine the first time the root bucket is accessed
+  /// via this getter. After the data has been copied over from the engine, this
+  /// getter will return a [SynchronousFuture], that immediately resolves to the
+  /// root [RestorationBucket].
+  ///
+  /// See also:
+  ///
+  ///  * [RootRestorationScope], which makes the root bucket available in the
+  ///    [Widget] tree.
+  Future<RestorationBucket> get rootBucket {
+    if (!_isListeningForEngineUpdates) {
+      SystemChannels.restoration.setMethodCallHandler(_methodHandler);
+      _isListeningForEngineUpdates = true;
+    }
+    if (_rootBucket != null) {
+      return SynchronousFuture<RestorationBucket>(_rootBucket);
+    }
+    if (_pendingRootBucket == null) {
+      _pendingRootBucket = Completer<RestorationBucket>();
+      _getRootBucketFromEngine();
+    }
+    return _pendingRootBucket.future;
+  }
+  RestorationBucket _rootBucket;
+  Completer<RestorationBucket> _pendingRootBucket;
+  bool _isListeningForEngineUpdates = false;
+
+  Future<void> _getRootBucketFromEngine() async {
+    final Map<String, dynamic> data = await retrieveFromEngine();
+    if (_pendingRootBucket == null) {
+      // The engine was faster in sending us the data via the 'push' method on
+      // the SystemChannel.
+      return;
+    }
+    assert(_rootBucket == null);
+    _setRootBucket(data);
+  }
+
+  void _setRootBucket(Map<String, dynamic> data) {
+    _rootBucket = _createRootBucket(data);
+    _pendingRootBucket?.complete(_rootBucket);
+    _pendingRootBucket = null;
+  }
+
+  RestorationBucket _createRootBucket(Map<String, dynamic> data) {
+    return RestorationBucket.root(manager: this, rawData: data);
+  }
+
+  Future<dynamic> _methodHandler(MethodCall call) {
+    switch (call.method) {
+      case 'push':
+        handleRestorationDataUpdate(call.arguments as Uint8List);
+        break;
+      default:
+        throw UnimplementedError("${call.method} was invoked but isn't implemented by $runtimeType");
+    }
+    return null;
+  }
+
+  /// Called by the [RestorationManager] on itself when the engine has provided
+  /// new serialized restoration `data`.
+  ///
+  /// The method decommissions the current [rootBucket] and replaces it with
+  /// the bucket hierarchy obtained by de-serializing the provided `data`.
+  @protected
+  void handleRestorationDataUpdate(Uint8List data) {
+    final RestorationBucket oldRoot = _rootBucket;
+    _setRootBucket(decodeRestorationData(data));
```

Comment:
e.g. here it's not clear to me if a subclass could/would ever call this method after the completer has completed.

Reply:
This is fine to call after the completer has completed. It will just replace the current `rootBucket` with the new one. If there's no completer, it just means that currently nobody is waiting for the root bucket.